climpred/bootstrap.py

```diff
@@ -1,14 +1,94 @@
 import inspect
+import multiprocessing
+import warnings
 
+import dask
 import numpy as np
 import xarray as xr
-from tqdm.auto import tqdm
 
 from .checks import has_dims
 from .constants import ALL_COMPARISONS, ALL_METRICS, METRIC_ALIASES
 from .prediction import compute_hindcast, compute_perfect_model, compute_persistence
 from .stats import dpp, varweighted_mean_period
-from .utils import assign_attrs, get_comparison_class, get_metric_class
+from .utils import (
+    _ensure_loaded,
+    _transpose_and_rechunk_to,
+    assign_attrs,
+    get_comparison_class,
+    get_metric_class,
+)
+
+ncpu = multiprocessing.cpu_count()
+
+
+def chunking_parallel_performance_increase_likely(ds, dim):
+    """Check whether chunking might make sense.
+
+    Criteria for potential performance increase:
+    - input xr.oject needs to be chunked realistically.
+    - input xr.object needs to sufficiently large so dask overhead doesn't
+     overcompensate parallel computation speedup.
+    - there should be several CPU available for the computation, like on a
+     cluster or multi-core computer
+    """
+    crit_size = 10000000
+    if not dask.is_dask_collection(ds):
+        if ds.size > crit_size and ncpu >= 4:
+            warnings.warn(
+                f'Consider chunking input ds along other dimensions than '
+                f'{dim} for parallelized performance increase.'
+            )
+    else:
+        if ds.size < crit_size:
+            warnings.warn(
+                'Chunking might not bring parallelized performance increase, '
+                f'because input size quite small, found ds.size = {ds.size} <'
+                f' {crit_size}.'
+            )
+        if ncpu < 4:
+            warnings.warn(
+                f'Chunking might not bring parallelized performance increase, '
+                f'because only few CPUs available, found {ncpu} CPUs.'
+            )
+        # if ds.chunks.count() much larger than nworkers, warn smaller chunks
+
+
+def resample(hind, shuffle_dim, to_be_shuffled):
+    """Resample hind in dimension `shuffle_dim` from values `to_be_shuffled`"""
+    smp = np.random.choice(to_be_shuffled, len(to_be_shuffled))
+    smp_hind = hind.sel({shuffle_dim: smp})
+    # ignore for resample init on ds in compute
+    if shuffle_dim != 'init':
+        smp_hind[shuffle_dim] = np.arange(1, 1 + smp_hind[shuffle_dim].size)
+    return smp_hind
+
+
+def _dask_percentile(arr, axis=0, q=95):
+    if len(arr.chunks[axis]) > 1:
+        msg = 'Input array cannot be chunked along the percentile ' 'dimension.'
+        raise ValueError(msg)
```

Comment:
Sorry I don’t understand your comment.

Reply:
Right now if you are applying the function over `axis` and it is chunked, it will break. This is the same as if you tried to do `rm_trend` but were chunked over `time` for instance.

Here you throw a hard error if this happens. I'm suggesting that instead you can do the same as we did in `compute_hindcast`. So if you notice there's multiple chunks over `axis` you just rechunk to `chunks = {'axis': -1}`. `-1` is the keyword to chunk over the full dimension.

Reply:
now this rechunks. performance stays.